text/0000-dyno.md

```diff
@@ -0,0 +1,294 @@
+
+`provide_value` and `provide_ref` are convenience methods for the common cases of providing a temporary value and providing a reference to a field of `self`, respectively. `provide_value` takes a function to avoid unnecessary work when querying for data of a different type; `provide_ref` does not use a function since creating a reference is typically cheap.
+
+`provide` and `provide_with` offer full generality, but require the explicit use of type tags.
+
+It seems reasonable that data might be accessed and provided on different threads. For this purpose, `provide_any` includes a version of `Requisition` which implements `Send`: `SendRequisition`. An open question is if it is also useful to support `Sync` variations (and if there is a better name).
+
+
+TODO docs from crate
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There is some overlap in use cases with `Any`. It is sub-optimal for the standard library to include two modules with such similar functionality. However, I believe `provide_any` is more of a complement to `any`, rather than an alternative: `any` supports type-hiding where the concrete type is chosen by the library, whereas with `provide_any`, the library user chooses the concrete type.
+
+`provide_any` could be a stand-alone crate, however, a key motivation for `provide_any` is use in `Error` and that is only possible if `provide_any` is part of the standard library.
+
+`provide_any` is fairly complex, however, this is mitigated by restricting the exposed complexity to intermediate implementers or to users with advanced use cases. For many Rust programmers, they won't even know `provide_any` exists, but will reap the benefits via more powerful error handling, etc. 
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are are few ways to do without the proposed feature: in some situations it might be possible to add concrete accessor methods to traits. A trait could implement generic member access based on a value identifier (i.e., name-driven rather than type driven), such a method would return a `dyn Any` trait object. Or a user could speculatively downcast a trait object to access its data.
+
+Each of these approaches has significant downsides: adding methods to traits leads to a confusing API and is impractical in many situations (including for `Error`). Value-driven access only works with types which implement `Any`, that excludes types with lifetimes and unsized types; furthermore it requires the caller to downcast the result which is error-prone and fragile. Speculative downcasting violates the encapsulation of trait objects and is only possible if all implementers are known (again, not possible with `Error`).
+
+`provide_any` could live in its own crate, rather than in libcore. However, this would not be useful for `Error`.
+
+`provide_any` could be a module inside `any` rather than a sibling (it could then be renamed to `provide` or `provider`).
```

Comment:
Living under `any` rather than in a new top-level module seems like a more natural home for this to me. There's overlap between type tags and type ids, and `Any` and `Provider` that I think we could give proper treatment together under the same umbrella.

Reply:
that makes sense! I'm a bit hesitant because type ids and type tags are so obviously similar but have nothing in common in their interfaces or how they are used, so I thought it might be a bit confusing to have them together